resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala

```diff
@@ -213,6 +216,14 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
       sc.conf.getOption("spark.mesos.driver.frameworkId").map(_ + suffix)
     )
 
+    // check that the credentials are defined, even though it's likely that auth would have failed
+    // already if you've made it this far, then start the token renewer
+    if (hadoopDelegationTokens.isDefined) {
```

Comment:
You shouldn't do this here, otherwise you need to keep that field `protected` in the parent class and that adds unnecessary coupling. Instead, do this in `initializeHadoopDelegationTokens`.

Reply:
I agree that I shouldn't need to use the conditional `hadoopDelegationTokens.isDefined`, however there will need to be some check (`UserGroupInformation.isSecurityEnabled` or similar) to pass the `driverEndpoint` to the renewer/manager here. When the initial tokens are generated `driverEndpoint` is still `None` because `start()` hasn't been called yet.

Reply:
I may have spoke too soon, there might be a way..

Reply:
You could call `initializeHadoopDelegationTokens` in `start` after everything that's needed is initialized. It would also better follow the scheduler's lifecycle.

Reply:
Check out the patch now. `hadoopDelegationTokens` now calls `initializeHadoopDelegationTokens` (renamed `fetchHadoopDelegationTokens`) by name:
```scala
  private val hadoopDelegationTokens: () => Option[Array[Byte]] = fetchHadoopDelegationTokens
```
 This has the effect of only generating the first set of delegation tokens once the first `RetrieveSparkAppConfig` message is received. At this point, everything has been initialized because renewer (renamed `MesosHadoopDelegationTokenManager`) is evaluated lazily with the correct `driverEndpoint`. 

It's a bit confusing to just avoid an extra conditional. WDYT?